### PR TITLE
Hyper-V provider overhaul

### LINKED ISSUE #8988: Hyper-V automatic checkpoints should default to disabled
Hyper-V added a new feature that automatically takes checkpoints of VMs when they're started. When you stop the VM, it's deleted. [Link](https://blogs.technet.microsoft.com/virtualization/2017/04/20/making-it-easier-to-revert/)

However - if you're trying to manage state with Vagrant this could get in the way. Personally I'd prefer to rely on `vagrant destroy` or `vagrant provision -force` to get back to a known-good state. What do you think?

I confirmed that it conflicts with Packer's post-processor for Vagrant https://github.com/hashicorp/packer/issues/5371 and can give errors like this

### Vagrant version
Vagrant 2.0.0

### Host operating system
`gin *BuildLabEx*` returns 16291.0.amd64fre.rs3_release.170912-1708
This would apply to any Windows 10 insider build since around May

### Guest operating system
Windows Server Insider, build 16278
[Packerfile](https://github.com/StefanScherer/packer-windows/blob/my/windows_2016_insider.json)

### Vagrantfile
```ruby
Vagrant.configure("2") do |config|
  config.vm.box = "rs3_release_16296_insider_hyperv"
end
```

### Expected behavior
VM should have started

### Actual behavior
```none
Bringing machine 'default' up with 'hyperv' provider...
==> default: Verifying Hyper-V is enabled...
==> default: Importing a Hyper-V instance
    default: Please choose a switch to attach to your Hyper-V instance.
    default: If none of these are appropriate, please open the Hyper-V manager
    default: to create a new virtual switch.
    default:
    default: 1) HvsiIcs
    default: 2) DockerNAT
    default: 3) nat
    default: 4) Ethernet
    default: 5) Default Switch
    default:
    default: What switch would you like to use? 4
    default: Cloning virtual hard drive...
    default: Creating and registering the VM...
An error occurred while executing a PowerShell script. This error
is shown below. Please read the error message and see if this is
a configuration error with your system. If it is not, then please
report a bug.

Script: import_vm_vmcx.ps1
Error:

"Missing Checkpoint present."
```

### Steps to reproduce

1. Build [Packerfile](https://github.com/StefanScherer/packer-windows/blob/my/windows_2016_insider.json)
1. `vagrant box add --name windows_2016_insider windows_2016_insider_hyperv.box`
1. `vagrant init windows_2016_insider`
1. `vagrant up`

### References

https://github.com/hashicorp/packer/issues/5371

### PR DESCRIPTION
**Pull request by chrisroberts**
This PR cleans up the Hyper-V provider. It removes old scripts that are no longer being used. The import process has been unified so there is only a single point of entry for both xml and vmcx based VMs. VM configuration has been extracted into its own action so guest modifications can be applied after creation. It also includes a bunch of cleanup, documenting, and test coverage.

Fixes #8988 #8877 #9317 #9540

### CODE REVIEW

**Comment by chrisroberts** [comment]
@PatrickLang @bbrala @StefanScherer @taliesins If any of you would like to review some/all of this PR and give any feedback, it would be greatly appreciated :smiley:

File: plugins/providers/hyperv/config.rb [hunk 1]
```diff
@@ -24,21 +55,20 @@ def initialize
...
-            time_synchronization: UNSET_VALUE,
-            vss: UNSET_VALUE
-        }
+        @enable_checkpoints = UNSET_VALUE
+        @vm_integration_services = {}
       end
 
       def finalize!
+        @linked_clone = false if @linked_clone == UNSET_VALUE
+        @differencing_disk = false if @differencing_disk == UNSET_VALUE
```
Comment (bbrala) [inline plugins/providers/hyperv/config.rb]: Wouldn't some sort of deprecation warning make sense somewhere around here? If it is deprecated, perhaps telling the user makes sense.

**Comment by bbrala** [comment]
I will take the time to take a good look some time soon, for now i (rather quicktly) read through the code, it does feel quite a bit more robust and clean. I think this will help debugging in the future :)

Anyhow, will be a bit more thorough and actually test the code soon.

File: lib/vagrant/util/powershell.rb [hunk 2]
```diff
@@ -220,6 +242,19 @@ def self.powerup_command(path, args, opts)
...
         instance_variables.each(&method(:remove_instance_variable))
       end
+
+      # @private
+      # This is a helper method that provides the PowerShell command to resize
+      # the "console" to prevent output wrapping or truncating. An environment
+      # variable guard is provided to disable the behavior in cases where it
+      # may cause unexpected results (VAGRANT_POWERSHELL_RESIZE_DISABLE)
+      def self.resize_console
+        if ENV["VAGRANT_POWERSHELL_RESIZE_DISABLE"]
```
Comment (briancain) [inline lib/vagrant/util/powershell.rb]: I think this setting should be documented so that it's clear it's available for use.

**Comment by bbrala** [comment]
@chrisroberts i know this has been merged, but i've done some basic tests on our default machine with some of the options and things seem to work without a problem.

**Pull request merged**